CppImport/src/macro/DefinitionManager.cpp

```diff
@@ -0,0 +1,118 @@
+ **      following disclaimer.
+ **    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+ **      following disclaimer in the documentation and/or other materials provided with the distribution.
+ **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "DefinitionManager.h"
+
+namespace CppImport {
+
+DefinitionManager::DefinitionManager(ClangHelper* clang) : clang_(clang) {}
+
+void DefinitionManager::addMacroDefinition(QString name, const clang::MacroDirective* md)
+{
+	definitions_[md] = name;
+}
+
+QString DefinitionManager::definitionName(const clang::MacroDirective* md)
+{
+	if (!definitions_.contains(md)) return nullptr;
+
+	return definitions_[md];
+}
+
+bool DefinitionManager::isPartialBegin(const clang::MacroDirective* md)
+{
+	return definitionName(md).startsWith("BEGIN_");
+}
+
+bool DefinitionManager::isPartialEnd(const clang::MacroDirective* md)
+{
+	return definitionName(md).startsWith("END_");
+}
+
+void DefinitionManager::clear()
+{
+	definitions_.clear();
+}
+
+bool DefinitionManager::macroDefinitionLocation(const clang::MacroDirective* md, QString& namespaceName,
+																QString& containerName)
+{
+	auto presumedLoc = clang_->sourceManager()->getPresumedLoc(md->getMacroInfo()->getDefinitionLoc());
+	auto path = QDir(presumedLoc.getFilename()).absolutePath();
+
+	/*
+	 * given a path: ../Envision/A/../B.xyz
+	 * this regex captures A (index 1) and B.xyz (index 3)
+	 */
+	QRegularExpression regex ("/Envision/(\\w+)(/.*/|/)(\\w+\\.\\w+)$", QRegularExpression::DotMatchesEverythingOption);
+	auto match = regex.match(path);
+	if (!match.hasMatch()) return false;
+
+	// use the directory name of A as the namespace name
+	namespaceName = match.captured(1);
+
+	// some Envision namespaces don't have the same name as the corresponding directory
+	if (namespaceName == "ModelBase")
```

Comment:
Could you please extract these mappings, into a static field, initialized at the beginning of the file? That way the functional code will be nicer and generic, only the map will be Envision specific.